guidesmiths/generator-systemic
# V2.3.x into master

### Description
This PR will add to the generator-systemic the possibility to generate not only the basic and without-business-logic version of a systemic-based microservice but as well the show-case version of it.

Once again with this approach we
- do not pollute the systemic-generator code and have to maintain 2 version of it
  - `yo systemic`
  - `yo systemic:advanced`
- do not have to bother the user with the useless and always NO answered question `Would you like to have a showcase including extra components?`

#### Usage:
- **basic version**: `yo systemic`
  - highlights the un-opinionated approach of systemic about how to create microservices. We give the users a base and they define which type of business logic they want to implement (rest + DB / bus + DB / bus + blob / rest + third party integration / etc.)
- **show-case version**: `yo systemic --extra-components`
  - lets the user understand how to get the most out of a systemic-based microservice.
  - fosters the implementation of an Hexagon Architecture for the microservice.
  - gives an hint of how to manage multiple business logic workflows in the same microservice.
  - gives an hint of how to manage versioning of components.
  - gives an hint of how to manage integration testing.
  - gives an hint of how to manage the infrastructure for a given microservice with docker-compose.
  - gives an hint of how big the systemic ecosystem is when it comes to already available npm systemic components (systemic-rabbitmq, systemic-mongodb).

### How to review this PR
The newly added `--extra-components` flag is retrieved in `generators/app/index.js` and then set as a prop for the generator. That prop is the one used in order to:
- add or not add a given file to the generated microservice. 
  - Check `props.filesToSkip` in `generators/app/index.js`
- add or not add a given bit of code to a file ending up being part of the generated microservice
  - e.g. `<%_ if (extraComponents) { -%> "error-handler-module": "^1.0.3", <%_ } -%>` in `generators/app/root/_package.json`

A more detailed description of this new capabilities can be found in:
- `README.md`
- `generators/app/templates/root/_README.md`

I would recommend to the reviewers to run `yo systemic --extra-components` and check the generated microservice starting from a carefully reading fo the generated README.


### Issues solved by this PR
- #17 --> the [error-handler-module package](https://github.com/guidesmiths/error-handler-module) has been added in the show-case version in order to show http errors handling capabilities. This would made the PR #41 obsolete.
- #54 
- #25 
- #59 
- #14 
- #11 
- #6
